pydepsi/network.py

```diff
@@ -0,0 +1,178 @@
+    # already has x arcs connected or already has all allowed arcs connected
+    # (e.g. there are no more nodes in that group, or they are all too far away).
+
+    arcs = []
+
+    indices = range(len(coordinates))
+    for cur_index in indices:
+        # Calculate per node, the group they are in and the distance from the current node.
+        groups = [
+            int(math.floor(num_groups * (0.5 + math.atan2(coordinate[1], coordinate[0]) / math.tau)))
+            for coordinate in coordinates - coordinates[cur_index]
+        ]
+        groups[cur_index] = num_groups + 1  # Separate the current node into its own group.
+        distances = [get_distance(coordinates[cur_index], coordinate) for coordinate in coordinates]
+
+        # Create a list of tuples with the group, distance, and index, sorted by group and then distance.
+        values = np.array(sorted(list(zip(groups, distances, indices, strict=False))))
+
+        # Collect the nearest max_links neighbors per group and discard the group of the current node.
+        groups_diff = values[1:, 0] - values[:-1, 0]
+        separators = np.where(groups_diff > 0)[0]
+        groups = np.split(values, separators + 1)
+        groups = groups[: len(groups) - 1]
+        groups = [group[:max_links] for group in groups]
+
+        # Collect the neighbor 'hierarchies'.
+        # Each hierarchy contains the nth nearest neighbor from all groups.
+        neighbor_hierarchies = [[] for _ in range(max_links)]
+        count = 0
+        for n in range(max_links):
+            # Break early if we have gathered enough neighbors.
+            if max_links <= count:
+                break
+            for group in groups:
+                # Note that we do not break inside this loop, because we want the nth nodes from all groups.
+                if n < len(group) and check_in_range(group[n][1], min_length, max_length):
+                    neighbor_hierarchies[n].append(group[n])
+                    count = count + 1
+        # Possibly, these loops could be replaced by this.
+        # neighbor_hierarchies = [np.stack([el for el in zipped \
+        #                         if (el is not None and check_in_range(el[1], min_length, max_length))]) \
+        #                         for zipped in itertools.zip_longest(*groups)]
+
+        # Sort hierarchies per group by distance to the current node.
+        neighbor_hierarchies = [
+            sorted(hierarchy, key=lambda x: x[1]) for hierarchy in neighbor_hierarchies if len(hierarchy) != 0
+        ]
+
+        # Add sorted arcs to at most max_links neighbors.
+        cur_arcs = [
+            tuple(sorted([cur_index, int(neighbor[2])])) for hierarchy in neighbor_hierarchies for neighbor in hierarchy
+        ]
+        cur_arcs = cur_arcs[:max_links]
+
+        arcs.extend(cur_arcs)
+
+    # TODO(tvl): Fix issue with max_links: a later node may add an extra arc to an already 'full' earlier node.
```

Comment:
As discussed, not needed. A point can be connected with more than min_links arcs.